### PR TITLE
Initial CAMARA OAS linting ruleset 

### LINKED ISSUE #15: Extend the API guidelines with a concrete ruleset
Several subprojects could be facing issues currently if API specification contributions are made without performing a linting check. To ensure that we can enable this process in Camara natively, we would need a concrete ruleset to validate the spec files against. This issue is created to propose an extension to the existing guidelines by adding a concrete ruleset.

An example of a subproject requesting such a check can be seen here https://github.com/camaraproject/QualityOnDemand/issues/158

### PR DESCRIPTION
**Pull request by rartych**
#### What type of PR is this?

Add one of the following kinds:
* enhancement/feature

#### What this PR does / why we need it:

The list of core Spectral OAS rules and proposed additional rules to be used for linting OAS files within subprojects

#### Which issue(s) this PR fixes:

Fixes #15

#### Special notes for reviewers:

Draft version for discussion, evolution and implementation details expected in the next round

#### Changelog input

```
Documentation for OAS linting ruleset added

```

#### Additional documentation 

[Spectral docs](https://docs.stoplight.io/docs/spectral/4dec24461f3af-open-api-rules)

### CODE REVIEW

File: documentation/Linting-rules.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+
+The full list of the rules in this ruleset are described in [OpenAPI Rules](https://docs.stoplight.io/docs/spectral/4dec24461f3af-open-api-rules).
+
+
+`extends: [[spectral:oas, off]]`  - this avoids running any rules from the extended ruleset as they are disabled. Each rule then can be [enabled individually](https://docs.stoplight.io/docs/spectral/0a73453054745-recommended-or-all#enabling-rules).
+
+### Rule severity
+
+The `severity` keyword is optional in rule definition and can be `error`, `warn`, `info`, `hint`, or `off`.
+The default value is `warn`.
```
Comment (Kevsy) [inline documentation/Linting-rules.md:20]: We should discuss what is the expected outcome for each severity. For example should `warn` result in a change and re-lint, or can it be ignored?

**Review by Kevsy — APPROVED** [review]

**Comment by Kevsy** [comment]
Excellent start @rartych - I 've submitted an Approval, and have made one question in the review about the expected outcome for each severity, which we can include n the next version.

File: documentation/Linting-rules.md (hunk 1 above)
Comment (hdamker) [reply documentation/Linting-rules.md:20]: Good point. Maybe we can differentiate along the following lines:

* warnings can be ignored when introducing linting first time
  * OR (alternative) warnings can be ignored if correcting them would introduce a breaking change
* warnings shall not be introduced with a change (e.g. within a merge, that's the point where linting is done)

But that's to clarify in a next version. Will you open an issue for that @Kevsy ?

File: documentation/Linting-rules.md [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+| `UE`                 | `device`           |
+| `MSISDN`                 | `phone number`           |
+| `mobile network`      | `network`         |
+
+
+_Spectral rule_: [camara-language-avoid-telco]()
+
+*Severity*: `hint`
+
+## 3. API Definition
```
Comment (sachinvodafone) [inline documentation/Linting-rules.md:102]: |  &nbsp; Topic|  &nbsp; Description |
|----------------------------|-------------------------|
| Openapi version                | Camara APIs MUST be based on OpenAPI specification with minimum Version 3.0.0.           |
| Request body structure on PUT, POST, PATCH    |    POST / PUT / PATCH methods must have a requestBody. The requestBody content must be "application/json".   |
| Info object: title, version, description   |  "Info" section is mandatory. Title, Description and Version must be documented.Note: It is provided by Spectral.    |
|  Info object: API version format  |   API version within the Info section must follow the format: X.Y.Z   |
| Valid basepath validation   |   API must have valid basepath   |
| Camel case / lower case /Snake_Case   |    Paths URI (tasks, individual resources, etc.) MUST be  lower case and hyphens. OperationIds must be defined in lowerCamelCase Objects must be defined in CamelCase inside properties field. The event type must be written in UPPER_SNAKE_CASE  |
| Path URI   |  URIs must contain the "major version" of the API.    |
| Resource path must not contain the base URL   | Path must not contain the base path.     |
| Sensitive data (msisdn/imsi) cannot be a path or query parameter   |     Sensitive data (msisdn/imsi) cannot be a path or query parameter.Note: Needs to list down if we have other sensitive parameters other than MSISDN/IMSI |
|Valid methods    |     Valid methods method are: GET, PUT, POST, DELETE, PATCH, OPTIONS, HEAD |
|  GET/DELETE methods must not accept a "body"  |  'GET' and 'DELETE' http methods MUST NOT accept a 'requestBody' attribute    |
| Define common error response pattern   |  Following error responses must be define: 400, 401, 403, 404, 405, 406, 429, 500, 501, 502, 503, 504.Note: List needs to be validated as per Camara API documentation    |
|  Verb methods cannot be in the Path  |   Resources must not contain the method name get, put, post, delete, patch.   |
| Parameters must have a description   |   All parameters must have a description.   |
| Path version must be major versions only   |  Path must contain the major API version with format /vX.Example: paths: "/device-status/v0/roaming"    |
|  GET/DELETE methods must not accept a "body"  |   'GET' and 'DELETE' http methods MUST NOT accept a 'requestBody' attribute   |
| Mandatory request body with POST, PATCH and PUT methods   |      The requestBody object MAY be present when the operation is used with one of PUT, POST, PATCH HTTP Verbs.|
|  Summary associated with each operation  |   Summary must be defined on each operation, describing with a short summary what the operation does.   |
| Include valid response examples   |   For each response code, we must define response payload examples   |

**Comment by sachinvodafone** [comment]
Hi @rartych , we can ignore duplicate items (16 & 17) that you have highlighted . It seems , some issue in copy/paste . Also I am aligned with point 6 (Event Type) and 10 (HEAD) comments.  

|  &nbsp; Topic|  &nbsp; Description |
|  |   |
|  |   |
|  |   |
|  |   |
|  |   |

**Comment by sachinvodafone** [comment]
Hi @rartych , we can ignore duplicate items (16 & 17) that you have highlighted . It seems , some issue in copy/paste . Also I am aligned with point 6 (Event Type) and 10 (HEAD) comments. Apart from this , few others validation can be considered:

|  &nbsp; Topic|  &nbsp; Description |
|----------------------------|-------------------------|
| 20. All schema properties should have a type.| |
|21. All schema should have properties | |
|22. Response body schema should use a valid Ref Object | The response body schema definition SHOULD use a Reference Object ($ref), MAY be defined inline as well.|
|23. requestBody must contain a schema with a valid ref object | The schema definition of the request body MUST use a Reference Object ($ref).|
| 24. Mandatory request header parameter "X-Correlator"|All Camara APIs MUST define at least the header parameter that is used to track the inbound requests and outbound responses.|

**Review by jlurien — COMMENTED** [review]
Very good starting point. Just some minor comments. the camara-* spectral rules are yet to be developed in a later iteration, right?

**Comment by rartych** [comment]
@jlurien Thank you for your comments.

> Just some minor comments. the camara-* spectral rules are yet to be developed in a later iteration, right?

We (within DT) are developing these additional rules (camara-*) and intend to extend this PR with the code. 
`Linting-rules.md` file is high-level documentation of the ruleset. We also need guidelines for use of linting in Subprojects (as Github Actions or locally). I prefer to have all parts in one PR, but we can split the task into more iterations starting from current definition of rules scopes. Any views on that?
 
Another question is testing the ruleset against current API specification in subprojects to minimize false positives.
Having rules implemented I would need the help of Subprojects maintainers to verify the rules by executing them off-line/locally before the ruleset is released for integration with workflows.

**Comment by jlurien** [comment]
> @jlurien Thank you for your comments.
> 
> > Just some minor comments. the camara-* spectral rules are yet to be developed in a later iteration, right?
> 
> We (within DT) are developing these additional rules (camara-*) and intend to extend this PR with the code. `Linting-rules.md` file is high-level documentation of the ruleset. We also need guidelines for use of linting in Subprojects (as Github Actions or locally). I prefer to have all parts in one PR, but we can split the task into more iterations starting from current definition of rules scopes. Any views on that?
> 
> Another question is testing the ruleset against current API specification in subprojects to minimize false positives. Having rules implemented I would need the help of Subprojects maintainers to verify the rules by executing them off-line/locally before the ruleset is released for integration with workflows.

My feeling is that this task is too big to keep all work in one PR, so we may iterate, but I let you decide what to include here. I can help testing with Device Location APIs

File: documentation/Linting-rules.md [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+### Rule severity
+
+The `severity` keyword is optional in rule definition and can be `error`, `warn`, `info`, `hint`, or `off`.
+The default value is `warn`.
+
+### OpenAPI v2 & v3
+Rules applying to both OpenAPI v2.0, v3.0, and most likely v3.1.
+
+
+|Name| Desc| Recom mended|CAMARA use|Spectral severity | CAMARA severity
```
[the remark this answers is not in the document]
Comment (rartych) [reply documentation/Linting-rules.md:30]: Clarification added in the introduction.

**Review by RubenBG7 — APPROVED** [review]

**Pull request merged**